xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -470,13 +482,77 @@ private void updateSelector() {
     private ServerRoutingConfig generateRoutingConfig(FilterChain filterChain) {
       HttpConnectionManager hcm = filterChain.getHttpConnectionManager();
       if (hcm.virtualHosts() != null) {
-        return ServerRoutingConfig.create(hcm.httpFilterConfigs(),
-            new AtomicReference<>(hcm.virtualHosts()));
+        AtomicReference<ImmutableMap<Route, ServerInterceptor>> interceptorRef =
+            new AtomicReference<>(generatePerRouteInterceptors(
+                hcm.httpFilterConfigs(), hcm.virtualHosts()));
+        return ServerRoutingConfig.create(new AtomicReference<>(hcm.virtualHosts()),
+            interceptorRef);
       } else {
         RouteDiscoveryState rds = routeDiscoveryStates.get(hcm.rdsName());
         checkNotNull(rds, "rds");
-        return ServerRoutingConfig.create(hcm.httpFilterConfigs(), rds.savedVirtualHosts);
+        AtomicReference<ImmutableMap<Route, ServerInterceptor>> interceptorRef =
```

Comment:
Why not store this in RouteDiscoveryState? There's no need to use a different AtomicReference for each FilterChain. Same as we do for `savedVirtualHosts`.

Reply:
but serverInterceptor is built from httpConfigs for each filterChain, not each `savedVirtualHosts`

Reply:
I see. Makes sense.